apps/hash-ai-worker-ts/src/activities/flow-activities/answer-question-action.ts

```diff
@@ -0,0 +1,448 @@
+  You are an expert data analyst. You will be provided with data to analyze, which may be in the form of text,
+  a graph of structured entities with links between them, a spreadsheet, or some combination of these.
+  
+  Your boss will ask you to answer a question based on the data provided to you. They might have asked for a specific output format,
+  e.g. 'Markdown', 'JSON', or 'CSV'. If it isn't specified, you use your best judgment.
+  
+  Your boss plans to use your answer to make a decision, so you make sure that it is accurate, concise, and clear.
+  If you can''t provide an answer based on the available data, you explain why, and request more data if it would help.
+  
+  You write Python to analyze any context provided, if you need it. In your code, you access the context from a file using the provided path.
+  Your Python code contains detailed comments about why each function is used, and how the data is accessed, including references to the shape of the context data.
+  You code complete the entire task in one file.
+  
+  You provide a confidence score with your answer, which is a number between 0 and 1. 
+  1 represents absolute certainty, and 0 a complete guess – you never provide answers with confidence 0, but instead explain why you can't answer.
+`);
+
+type ModelResponseArgs = {
+  answer?: string;
+  explanation: string;
+  confidence?: number;
+  code?: string;
+};
+
+const callModel = async (
+  messages: OpenAI.ChatCompletionCreateParams["messages"],
+  context: string | null,
+  codeUsed: string | null,
+): Promise<
+  Status<{
+    outputs: StepOutput[];
+  }>
+> => {
+  const openApiPayload: OpenAI.ChatCompletionCreateParams = {
+    model: "gpt-4-turbo-preview",
+    messages,
+    stream: false,
+    temperature: 0,
+    tools: answerTools,
+  };
+
+  const openAiResponse = await getOpenAiResponse(openApiPayload);
+
+  logger.debug(`Open AI Response received: ${stringify(openAiResponse)}`);
+
+  const { response } = openAiResponse.contents[0]!;
+
+  const { message: modelResponseMessage } = response;
+
+  const toolCalls = modelResponseMessage.tool_calls;
+
+  if (!toolCalls?.length) {
+    return {
+      code: StatusCode.Internal,
+      message: "No tool call found in response",
+      contents: [],
+    };
+  }
+
+  const responseMessages: ChatCompletionToolMessageParam[] = [];
+
+  for (const toolCall of toolCalls) {
+    let parsedArguments: ModelResponseArgs;
+    try {
+      parsedArguments = JSON.parse(
+        toolCall.function.arguments,
+      ) as ModelResponseArgs;
+    } catch (err) {
+      return callModel(
+        [
+          ...messages,
+          modelResponseMessage,
+          {
+            role: "tool",
+            content: `Your JSON arguments could not be parsed – the parsing function errored: ${
+              (err as Error).message
+            }. Please try again.`,
+            tool_call_id: toolCall.id,
+          },
+        ],
+        context,
+        null,
+      );
```

Comment:
Does this risk an infinite loop, since we aren't using some `retryCount` parameter? I suppose we'll eventually hit the context limit, but might be better not to rely on that.

Separately, just to flag this approach could result in missed tool calls if the model made one correct tool call and one incorrect tool call. But not worth worrying about if we only have two tools for the call anyways.

Reply:
It does – fixed in https://github.com/hashintel/hash/pull/4290/commits/61e269b58f86ba46ac83bd50f876535c406c840c